### PR TITLE
Beginning OOP with Ruby

### PR DESCRIPTION
**Pull request by z64**
I wrote this a long time ago, currently living in my Gist here: https://gist.github.com/z64/343d8c1dfcac593f1bcfbb7b6f63d962

Todo:
- [x] Make links actual links.
- [x] ~~Our inline code blocks kind of look like shit. Maybe improve this?~~ (Thanks @unleashy !)
- [x] Remove Gist post with notice that it's been moved to the blog

### CODE REVIEW

File: _posts/2017-07-29-ruby-oop.md [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+To accomplish this, we will be solving a simple geometry problem:
+
+- Describe a collection of objects that can calculate the perimeter of a polygon with an arbitrary number of sides.
+
+## Breaking it down
+
+The first thing we want to do to tackle this problem is to break apart the problem into a series of smaller, bite sized questions. "Good Ruby code" will read a lot like english - I think you'll find the syntax we end up writing will almost look like english answers to these questions. You probably already know the answers to these (which is why I picked this as an example), but read them and pay attention to the language that I use.
+
+**1) What is perimeter?**
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md]: Maybe this should be "what is **a** perimeter"?

File: _posts/2017-07-29-ruby-oop.md [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+This post attempts to guide a beginner through object-oriented progamming in Ruby, showing them the thought process behind both designing and interacting with a library, and problem solving in general with Ruby.
+
+To accomplish this, we will be solving a simple geometry problem:
+
+- Describe a collection of objects that can calculate the perimeter of a polygon with an arbitrary number of sides.
+
+## Breaking it down
+
+The first thing we want to do to tackle this problem is to break apart the problem into a series of smaller, bite sized questions. "Good Ruby code" will read a lot like english - I think you'll find the syntax we end up writing will almost look like english answers to these questions. You probably already know the answers to these (which is why I picked this as an example), but read them and pay attention to the language that I use.
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md]: Capitalize "english" you fuck

File: _posts/2017-07-29-ruby-oop.md [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+> A **polygon** is an object **made up of lines**.
+
+**3) Okay, so what is a line?**
+
+> A line is the shortest distance between **two points** in space.
+Since we're dealing with finite lines, they have a *start* and an *end*.
+
+**4) Okay, so what is a point?**
+
+> A point is a location in space. For our purposes, a location on a 2D plane, bound by **coordinates**.
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md]: I think "For our purposes, this location is on a 2D plane, bounded by **coordinates**." sounds better

File: _posts/2017-07-29-ruby-oop.md [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+# A collection of lines that describe a 2D shape.
+class Polygon
+end
+```
+And from here, we can already describe how we're going to solve our problem, by creating a `#new` instance of a `Polygon`, and calling `#perimeter` on it:
+```ruby
+Polygon.new(___).perimeter
+```
+> In Ruby, and in most OOP languages, a `class` describes a "template" for an object. An object is usually a collection of data, and a set of methods or functions that allow you to interact with that data or give it logic. We first describe the objects with a `class`, and create new instances of them with `Thing.new`.
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md:59]: `Thing.new` feels broad, maybe say like "`X.new`, where X is the name of your class"? Not sure

File: _posts/2017-07-29-ruby-oop.md [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+```
+> `def` defines a new method, or an instruction on how to do something with an object. `initialize` is a special method that is called whenever you call `#new` on an object to make a new instance of one.
+
+We can now make a new `Point` like so:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.inspect #=> #<Point:0x2ae37f0 @x=1.5, @y=-3.0>
+```
+
+We're going to need to use these attributes of a point in order to do our calculations. To do so, we can define methods that will return the values of our attributes:
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md]: I think you need to explain why instance variables can't be accessed from outside, and why these methods are necessary. dunno

File: _posts/2017-07-29-ruby-oop.md [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+Playing with our new readers:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.x #=> 1.5
+puts point.y #=> -3.0
+```
+
+What else would be helpful to do with our point, or perhaps any other point? Well, with any two points, you can calculate the distance between them.
+
+Ruby has a rich standard library, of which includes mathematical functions. Needing the pythagorean theorem here, if we do some searching we can see Ruby already provides a method for us to calculate this instead of having to implement it ourselves:
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md]: "Needing the pythagorean theorem here, if we do some searching we can see ..."
Kinda tangled my brain a little. Something like "We need the Pythagorean theorem to calculate the distance between two points, and doing some searching we can see ..." would work better

File: _posts/2017-07-29-ruby-oop.md [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+  # Calculates the distance to another Point object
+  def distance(other)
+    x_difference = x - other.x
+    y_difference = y - other.y
+
+    Math.hypot x_difference, y_difference
+  end
+end
+```
+> Here, the `x` in `x_difference = x - other.x` is actually calling this objects *own* `attr_reader` method internally that fetches its `@x` attribute. In another language, you may write something like `self.x` to refer to this value. You can do that here too in Ruby, but in most cases you can omit `self.`, as it is already understood what you mean.
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md]: "..., but in most cases you can omit `self.`, as it is already understood what you mean."
Something like "but in most cases you can omit `self.`, because Ruby understands what you mean" if that doesn't sound cringy as fuck to you

File: _posts/2017-07-29-ruby-oop.md [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+  attr_reader :segments
+
+  def initialize(segments = [])
+    @segments = segments
+  end
+end
+```
+> Arguments for methods can be given a default value using the above syntax in `def initialize`. Here, I'm making `segments` default to an empty array.
+
+We're going to design `Polygon` to accept an `Array` of `Line`.
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md:253]: You need to explain what an array is, maybe? /shrug

File: _posts/2017-07-29-ruby-oop.md [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+Our new `Polygon` can be instanced like so..
```
Comment (unleashy) [inline _posts/2017-07-29-ruby-oop.md]: Instantiated *

**Review by unleashy — CHANGES_REQUESTED** [review]
lambasted you

File: _posts/2017-07-29-ruby-oop.md (hunk 1 above)
Comment (z64) [reply _posts/2017-07-29-ruby-oop.md]: No, this is referring to perimeter as a concept, not a particular perimeter.

File: _posts/2017-07-29-ruby-oop.md (hunk 4 above)
Comment (z64) [reply _posts/2017-07-29-ruby-oop.md:59]: It's entirely meant to be broad. But in any case, I wouldn't chose `X` as the name might confuse with the concepts being presented in the tutorial. (X, Y coordinates)

File: _posts/2017-07-29-ruby-oop.md (hunk 5 above)
Comment (z64) [reply _posts/2017-07-29-ruby-oop.md]: Tricky without getting too off topic. I'll try.

File: _posts/2017-07-29-ruby-oop.md (hunk 8 above)
Comment (z64) [reply _posts/2017-07-29-ruby-oop.md:253]: I think that may be out of the scope of the tutorial (basic, high level OOP design) but I can add a small note. (i.e., this isn't a *beginner's Ruby* tutorial, it's an OOP tutorial)

Comment (unleashy) [reply _posts/2017-07-29-ruby-oop.md:253]: O fuck right. I thought this was a beginner tutorial thing. aight

**Review by unleashy — APPROVED** [review]

**Comment by z64** [comment]
@Daniel-Worrall I plan to have this published by Friday. I'll accept edits after publishing, but it would be preferred to incorporate any of your suggestions beforehand if you can get around to it by then :+1: 

image

**Comment by unleashy** [comment]
R U B Y

T A C O

File: _posts/2017-07-29-ruby-oop.md [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+layout: post
+author: z64
+title: Beginning OOP with Ruby
+date: 2017-07-29
+tags: ruby oop tutorial
+---
+
+This post attempts to guide a beginner through object-oriented progamming in Ruby, showing them the thought process behind both designing and interacting with a library, and problem solving in general with Ruby.
+
+To accomplish this, we will be solving a simple geometry problem:
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md]: Probably reword this to "we will solve" to sound more like the intention of the post, rather than a plan to solve.

Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md]: Probably reword this to "we will solve" to sound more like the intention of the post, rather than a plan to solve.

File: _posts/2017-07-29-ruby-oop.md [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+
+## Breaking it down
+
+The first thing we want to do to tackle this problem is to break apart the problem into a series of smaller, bite sized questions. "Good Ruby code" will read a lot like English - I think you'll find the syntax we end up writing will almost look like English answers to these questions. You probably already know the answers to these (which is why I picked this as an example), but read them and pay attention to the language that I use.
+
+**1) What is perimeter?**
+
+> Perimeter is the total distance around an object, in our case a 2D polygon with any number of sides.
+
+**2) What is a Polygon?**
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md]: for all 4 of these points, we should keep consistent about capitalising the subjects. Either lowercase as speaking about them in general cases, or uppercase as preparation to refer to them as a class object. Also, we can use inline codeblocks if that is the intention.

File: _posts/2017-07-29-ruby-oop.md [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+**3) Okay, so what is a line?**
+
+> A line is the shortest distance between **two points** in space.
+Since we're dealing with finite lines, they have a *start* and an *end*.
+
+**4) Okay, so what is a point?**
+
+> A point is a location in space. For our purposes, this location is on a 2D plane, bound by **coordinates**.
+
+In the next section, we'll convert these concepts into Ruby code. Please re-read this section once we're done and see how what we said here is directly represented in the answer we come up with.
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md]: Remove the please to make it more direct, and possibly add a part at the bottom that says to go back up and read this part. Maybe with a link too. Or move this second part to the end of the post.

File: _posts/2017-07-29-ruby-oop.md [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+In the next sections, we'll work with one approach to solve this from the ground up, to figure out how to *make* a `Polygon` to get what we need, and explore some properties of Ruby along the way.
+
+## Objects
+
+### Point
+
+Consider any kind of object around you. A pen, a car, your computer - they all have properties, or *attributes*.  Ruby objects can be given *attributes* that let you describe and store data inside them. We do this with the `@var` syntax, like so, describing what defines a `Point` in 2D space:
+
+```ruby
+# A location in 2D space.
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md:70]: Possibly call this position instead of location

File: _posts/2017-07-29-ruby-oop.md [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+```
+> In our case, we will only need to ever read these attributes, so I just used `attr_reader`. `attr_accessor` would have also worked here.
+
+Playing with our new readers:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.x #=> 1.5
+puts point.y #=> -3.0
+```
+> You can read more about attributes, readers, setters, and more [here](http://ruby-for-beginners.rubymonstas.org/writing_classes/instance_variables.html).
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md:133]: Perhaps move this up between where you described what they are and when you use them.

File: _posts/2017-07-29-ruby-oop.md [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+
+What else would be helpful to do with our point, or perhaps any other point? Well, with any two points, you can calculate the distance between them.
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md]: I like how it's posed as a question, but this comes across very wordy. Could change to "What could we add to our `Point`? With two points, we can calculate the distance between them."

File: _posts/2017-07-29-ruby-oop.md [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+  # Calculates the distance to another Point object
+  def distance(other)
+    x_difference = x - other.x
+    y_difference = y - other.y
+
+    Math.hypot x_difference, y_difference
+  end
+end
+```
+> Here, the `x` in `x_difference = x - other.x` is actually calling this objects *own* `attr_reader` method internally that fetches its `@x` attribute. In another language, you may write something like `self.x` to refer to this value. You can do that here too in Ruby, but in most cases you can omit `self.`, because Ruby understands what you mean.
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md:157]: remove "actually", then change second case of "Ruby" to "it" or similar. Could give an example of another language to give the user a reference to what other languages you could be referring to if Ruby is their first language.

File: _posts/2017-07-29-ruby-oop.md [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+
+Let's refer back to our definition of a `Line`.
+
+```ruby
+# A finite segment (or distance) between two points.
+class Line
+end
+```
+
+From what we know, we can now give our `Line`s two attributes. We'll call them `start` and `stop` (I would use the word `end` here, but this is a Ruby keyword and it could get confusing.)
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md]: Maybe add to say that you cannot use Ruby keywords as variable names, not that it's confusing to use. (Yes, there are symbol workarounds, but I think it's better to talk about this rather than leave it vague)

File: _posts/2017-07-29-ruby-oop.md [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+
+Great. What other properties does a finite line have? Ah, right - length.
+
+But - we've already implemented this! We already know the distance between any two `Point` objects.. let's use it:
+
+```ruby
+# A finite segment (or distance) between two points.
+class Line
+  def length
+    start.distance stop
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md]: Explain how this line relates to the earlier code and that start and stop are points

File: _posts/2017-07-29-ruby-oop.md [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+
+## Moving forward
+
+There's tons of cool things out there that you can do with Ruby, and maybe even problems in your own life that you can use to cook up Ruby solutions for.
+
+Check out [http://awesome-ruby.com/](http://awesome-ruby.com/) for *tons* of community created packages - see what you can make with them!
+
+Learn to read documentation like those on [http://www.rubydoc.info/](http://awesome-ruby.com/). Then you'll be a real library ninja, and can answer a lot of your own questions! (Hopefully!)
+
+When you're stuck, there's a lot of helpful ruby people out there. StackOverflow, IRC, etc. - or even me! I can be reached at wherever you're viewing this post, or at [my email](mailto:zachnowicki@gmail.com), or on [Discord](https://discordapp.com) - I'm `z64#2639`.
```
Comment (Daniel-Worrall) [inline _posts/2017-07-29-ruby-oop.md:324]: Could also add a bit about the blog.

**Review by Daniel-Worrall — APPROVED** [review]
Overall, decent write up. Just some style / language change suggestions. Will set this to approve as none of them are critical, but should at least be looked over.

File: _posts/2017-07-29-ruby-oop.md (hunk 16 above)
Comment (z64) [reply _posts/2017-07-29-ruby-oop.md:157]: By saying "actually" here I'm attempting to distinguish it as a method call, with the assumption the reader might think it's just another temporary (local) variable.

Beyond this I'd rather not dive into the tangent of how other language's method / function scopes work.

File: _posts/2017-07-29-ruby-oop.md (hunk 13 above)
Comment (z64) [reply _posts/2017-07-29-ruby-oop.md:70]: I refer to it as a location in the previous section. I think this semantic is fine.

File: _posts/2017-07-29-ruby-oop.md (hunk 14 above)
Comment (z64) [reply _posts/2017-07-29-ruby-oop.md:133]: I think this is an appropriate place for this footer, as linking it earlier may distract someone from what I'm explaining if they open a link to the other site. Here, I've fully demonstrated attributes and getters/setters, and am about to move on to another topic ; this is a safe place to deviate from the lesson if someone wants to know more about them.
